### PR TITLE
feat: validate and trim SUPABASE_API_ENDPOINT

### PR DESCRIPTION
**Pull request by bhusareMayur**
## What does this PR do?
This PR adds validation and whitespace trimming to the `endpoint` configuration and `SUPABASE_API_ENDPOINT` environment variable during provider configuration. 

## Why is this necessary?
Previously, if a user accidentally included a trailing space in their environment variable (e.g., `export SUPABASE_API_ENDPOINT="https://api.supabase.com "`) or provided a malformed URL without a scheme, the raw string was passed directly to the HTTP client. This resulted in cryptic, low-level network errors during `terraform plan` or `apply`. 

By validating the endpoint upfront using `net/url`, we can catch human error instantly and return a helpful, actionable Terraform Diagnostic error.

## Changes
* Added `strings.TrimSpace` to the resolved `apiEndpoint`.
* Added `url.ParseRequestURI` validation to ensure the endpoint has a valid structure and an `http`/`https` scheme.
* Emits an explicit `AddAttributeError` diagnostic if validation fails.
* Added table-driven acceptance tests in `provider_test.go` to prove invalid inputs are caught and padded inputs are trimmed successfully.

## Demo / Terminal Output

**Before this PR (Cryptic Network Error):**
`Error: querying branches: Get "api.supabase.com/v1/projects/...": unsupported protocol scheme ""`

**After this PR (Clear Diagnostic Error):**
`Error: Invalid Supabase API Endpoint`

  `with provider["registry.terraform.io/supabase/supabase"],`
  `on main.tf line 1, in provider "supabase":`
   `1: provider "supabase" {`

`The endpoint "api.supabase.com" is not a valid URL. It must be a fully-qualified HTTP/HTTPS URL, for example: https://api.supabase.com`

### CODE REVIEW

**Comment by bhusareMayur** [comment]
Hi team,

I am opening this PR to make the API endpoint configuration more robust and to patch an edge case with the context cancellation in the HTTP retry transport (retryGETTransport.RoundTrip).
All CI checks have passed successfully. Please let me know if you would like any changes to the implementation or the gock tests. I am happy to iterate on this.

File: internal/provider/provider.go [hunk 1]
```diff
@@ -113,12 +115,27 @@ type retryGETTransport struct {
...
+
+	// If the request context was cancelled while the underlying transport
+	// returned a response, we must treat that as a cancellation and not
+	// return the response to callers. Close the body to avoid leaks.
+	if err == nil && req.Context().Err() != nil {
+		if resp != nil && resp.Body != nil {
+			_ = resp.Body.Close()
+		}
+		return nil, req.Context().Err()
 	}
```
Comment (JEETDESAI25) [inline internal/provider/provider.go]: Please remove this change and restore the previous `RoundTrip` implementation.

PR #309 intentionally makes `getOnlyRetryPolicy` return `(false, nil)` when a response exists, even if the context was cancelled. This lets callers inspect the response instead of losing it to `ctx.Err()`. The code comment and the test `"context canceled with existing response returns no error"` document this behavior. The new block closes the response and undoes that fix.

This also affects non-GET requests. If the server applies a write but cancellation happens as the response returns, Terraform may report failure and its state could become out of sync.

This transport change is unrelated to endpoint validation. No test covers this response-present path through `RoundTrip`, which is why CI still passes.

File: internal/provider/provider.go [hunk 2]
```diff
@@ -188,8 +205,23 @@ func (p *SupabaseProvider) Configure(ctx context.Context, req provider.Configure
...
 	if apiEndpoint == "" {
 		apiEndpoint = defaultApiEndpoint
+	} else {
+		// Validate URL structure, scheme, and host
+		u, err := url.ParseRequestURI(apiEndpoint)
+		if err != nil || (u.Scheme != "https" && u.Scheme != "http") || u.Host == "" {
+			resp.Diagnostics.AddAttributeError(
+				path.Root("endpoint"),
+				"Invalid Supabase API Endpoint",
+				fmt.Sprintf("The endpoint %q is not a valid URL. It must be a fully-qualified HTTP/HTTPS URL, for example: https://api.supabase.com", apiEndpoint),
```
Comment (JEETDESAI25) [inline internal/provider/provider.go]: Please strengthen this validation. The current condition still accepts endpoints that the generated client cannot use safely.

**Check the hostname.** `url.ParseRequestURI("https://:443")` succeeds because `u.Host` is `":443"`, even though there is no hostname. Use `u.Hostname() == ""` so port-only endpoints are rejected.

**Use `url.Parse`.** `ParseRequestURI` is intended for HTTP request targets and does not handle fragments correctly here. For example, it parses `https://gateway.example/supabase#frag` with `#frag` inside `Path` and leaves `Fragment` empty. Use `url.Parse` for this configuration URL.

**Reject queries and fragments.** The generated client appends `/` to the endpoint and resolves API paths relative to it. An endpoint such as `https://gateway.example/supabase?tenant=x` can silently become `https://gateway.example/v1/...`, losing both the base path and query. Fragment components and a bare `?` have the same problem, so check `RawQuery`, `ForceQuery`, and `Fragment`.

**Validate explicit ports.** `https://localhost:99999` passes URL parsing but fails later with `address 99999: invalid port`. If a port is provided, require it to be between 1 and 65535.

A possible implementation is:

```go
u, err := url.Parse(apiEndpoint)
invalid := err != nil ||
    (u.Scheme != "https" && u.Scheme != "http") ||
    u.Hostname() == "" ||
    u.RawQuery != "" || u.ForceQuery || u.Fragment != ""

if !invalid {
    if portText := u.Port(); portText != "" {
        port, portErr := strconv.Atoi(portText)
        invalid = portErr != nil || port < 1 || port > 65535
    }
}
```

This requires adding `strconv` to the imports. Using `url.Parse` still rejects the existing invalid cases: `api.supabase.com` has no scheme, and `http:////api.supabase.com` has no hostname.

Please also update the diagnostic message. A URL containing a query or fragment can be well-formed but unsupported as an API endpoint, so saying it is “not a valid URL” is misleading. A clearer message would be:

```text
The endpoint %q is not a valid Supabase API endpoint. Use a fully-qualified HTTP or HTTPS URL with a hostname, no query or fragment, and, if specified, a port from 1 to 65535.
```

File: internal/provider/provider.go [hunk 3]
```diff
@@ -188,8 +205,23 @@ func (p *SupabaseProvider) Configure(ctx context.Context, req provider.Configure
...
+			)
+			return
```
Comment (JEETDESAI25) [inline internal/provider/provider.go]: Please remove this `return`.  Configure normally collects all configuration errors before stopping at the `resp.Diagnostics.HasError()` check on line 247. This early return skips access token validation.

For example, if the endpoint is invalid and the access token is missing, the user sees only the endpoint error. After fixing it and running Terraform again, they then see the token error. Removing this return lets both errors appear together, while the existing check still prevents client creation.

File: internal/provider/provider.go [hunk 4]
```diff
@@ -188,8 +205,23 @@ func (p *SupabaseProvider) Configure(ctx context.Context, req provider.Configure
 	if !data.Endpoint.IsNull() {
 		apiEndpoint = data.Endpoint.ValueString()
 	}
+
+	// Trim whitespace from the endpoint
+	apiEndpoint = strings.TrimSpace(apiEndpoint)
+
 	if apiEndpoint == "" {
 		apiEndpoint = defaultApiEndpoint
```
Comment (JEETDESAI25) [inline internal/provider/provider.go]: Please keep trimming valid endpoints, but handle whitespace only values separately. Previously, `SUPABASE_API_ENDPOINT=" "` failed instead of making a request. With this change, it becomes empty and silently uses the production endpoint, `https://api.supabase.com`. A misconfigured self-hosted deployment could therefore send its access token to production.

Please keep the original value before trimming. If it is non-empty but becomes empty after `TrimSpace`, report an invalid-endpoint diagnostic. An exactly empty value should continue to use the default endpoint.

File: internal/provider/provider_test.go [hunk 5]
```diff
@@ -322,3 +322,62 @@ data "supabase_branch" "test" {
 		},
 	})
 }
+
+func TestAccProviderConfigure_EndpointValidation_Invalid(t *testing.T) {
+	// Verify that invalid endpoints trigger a diagnostic error and halt execution.
+	invalidEndpoints := []string{
+		"api.supabase.com",          // Missing scheme
+		"http:////api.supabase.com", // Malformed
+	}
```
Comment (JEETDESAI25) [inline internal/provider/provider_test.go:340]: Please add regression tests for these invalid endpoint cases:

```go
"ftp://api.supabase.com",                    // Unsupported scheme
"https://:443",                              // Port without hostname
"https://localhost:99999",                   // Port out of range
"https://gateway.example/supabase?tenant=x", // Query component
"https://gateway.example/supabase?",         // Empty query component
"https://gateway.example/supabase#frag",     // Fragment component
" ",                                         // Whitespace-only endpoint
```

**Review by JEETDESAI25 — COMMENTED** [review]
Thanks for the contribution @bhusareMayur . Endpoint validation is a useful addition, and trimming surrounding whitespace from valid endpoints is the right approach. Please address the inline comments before this is ready to merge.

**Comment by bhusareMayur** [comment]
@JEETDESAI25  Thank you for the incredibly thorough review and for catching those edge cases! I have addressed all of your feedback:

* Reverted the `RoundTrip` implementation to its original state to prevent any state drift.
* Updated the validation to use `url.Parse`, added strict checks for ports, queries, and fragments, and updated the diagnostic message.
* Removed the early return so that both endpoint and token configuration errors can surface simultaneously.
* Added logic to capture the original endpoint string to ensure whitespace-only inputs throw an error instead of defaulting to the production API.
* Added all requested regression test cases to `provider_test.go`.

All tests are passing locally with `TF_ACC="1"`. Please let me know if anything else is needed!

File: internal/provider/provider.go [hunk 6]
```diff
@@ -206,21 +191,43 @@ func (p *SupabaseProvider) Configure(ctx context.Context, req provider.Configure
...
 	} else {
-		// Validate URL structure, scheme, and host
-		u, err := url.ParseRequestURI(apiEndpoint)
-		if err != nil || (u.Scheme != "https" && u.Scheme != "http") || u.Host == "" {
+		// Validate URL structure, scheme, host, queries, fragments, and ports
+		u, err := url.Parse(apiEndpoint)
+		invalid := err != nil ||
+			(u.Scheme != "https" && u.Scheme != "http") ||
+			u.Hostname() == "" ||
+			u.RawQuery != "" || u.ForceQuery || u.Fragment != ""
```
Comment (JEETDESAI25) [inline internal/provider/provider.go]: One fragment edge case was missing from the implementation suggested. `url.Parse` does not preserve a trailing bare `#`, so `u.Fragment` remains empty and an endpoint such as `https://gateway.example/supabase#` passes validation. The generated client then drops the `/supabase` base path.

Checking the original string catches both empty and non empty fragments. This suggestion also rejects an explicitly empty port for consistency with the diagnostic requirement that ports be between 1 and 65535.

```suggestion
		invalid := err != nil ||
			(u.Scheme != "https" && u.Scheme != "http") ||
			u.Hostname() == "" ||
			u.RawQuery != "" || u.ForceQuery ||
			strings.Contains(apiEndpoint, "#") ||
			strings.HasSuffix(u.Host, ":")
```

File: internal/provider/provider_test.go [hunk 7]
```diff
@@ -326,8 +326,15 @@ data "supabase_branch" "test" {
...
-		"http:////api.supabase.com", // Malformed
+		"api.supabase.com",                          // Missing scheme
+		"http:////api.supabase.com",                 // Malformed
+		"ftp://api.supabase.com",                    // Unsupported scheme
+		"https://:443",                              // Port without hostname
+		"https://localhost:99999",                   // Port out of range
+		"https://gateway.example/supabase?tenant=x", // Query component
+		"https://gateway.example/supabase?",         // Empty query component
+		"https://gateway.example/supabase#frag",     // Fragment component
+		"   ",                                       // Whitespace-only endpoint
```
Comment (JEETDESAI25) [inline internal/provider/provider_test.go:339]: Please add regression coverage for the empty fragment and empty port cases while keeping the existing whitespace only case.

```suggestion
		"https://gateway.example/supabase#",         // Empty fragment
		"https://selfhosted.example.com:",           // Empty explicit port
		"   ",                                       // Whitespace-only endpoint
```

**Review by JEETDESAI25 — COMMENTED** [review]
Thanks for the thorough turnaround @bhusareMayur. Every point from the original review has been addressed correctly, and the improved diagnostic wording is a nice touch.

One fragment edge case was missing from the implementation suggested. I have left two small committable suggestions to close that case.

Please apply both suggestions and this will be ready to approve from my side.

**Comment by bhusareMayur** [comment]
@JEETDESAI25  Thank you for the review and for catching those final edge cases. I have applied both suggestions and verified that all tests continue to pass locally. Everything should be ready to go!

**Review by JEETDESAI25 — APPROVED** [review]
